src/category_theory/abelian/pseudoelements.lean

```diff
@@ -0,0 +1,411 @@
+lemma.
+
+Pseudoelements are in some ways weaker than actual elements in a concrete category. The most
+important limitation is that there is no existensionality principle: If `f g : X ⟶ Y`, then
+`∀ x ∈ X, f x = g x` does not necessarily imply that `f = g` (however, if `f = 0` or `g = 0`,
+it does). A corollary of this is that we can not define arrows in abelian categories by dictating
+their action on pseudoelements. Thus, a usual style of proofs in abelian categories is this:
+First, we construct some morphism using universal properties, and then we use diagram chasing
+of pseudoelements to verify that is has some desirable property such as exactness.
+
+It should be noted that the Freyd-Mitchell embedding theorem gives a vastly stronger notion of
+pseudoelement (in particular one that gives existensionality). However, this theorem is quite
+difficult to prove and probably out of reach for a formal proof for the time being.
+
+## Main results
+
+We define the type of pseudoelements of an object and, in particular, the zero pseudoelement.
+
+We prove that every morphism maps the zero pseudoelement to the zero pseudoelement (`apply_zero`)
+and that a zero morphism maps every pseudoelement to the zero pseudoelement (`zero_apply`)
+
+Here are the metatheorems we provide:
+* A morphism `f` is zero if and only if it is the zero function on pseudoelements.
+* A morphism `f` is an epimorphism if and only if it is surjective on pseudoelements.
+* A morphism `f` is a monomorphism if and only if it is injective on pseudoelements
+  if and only if `∀ a, f a = 0 → f = 0`.
+* A sequence `f, g` of morphisms is exact if and only if
+  `∀ a, g (f a) = 0` and `∀ b, g b = 0 → ∃ a, f a = b`.
+* If `f` is a morphism and `a, a'` are such that `f a = f a'`, then there is some
+  pseudoelement `a''` such that `f a'' = 0` and for every `g` we have
+  `g a' = 0 → g a = g a''`. We can think of `a''` as `a - a'`, but don't get too carried away
+  by that: Pseudoelements of an object to not form an abelian group.
+
+## Notations
+
+We introduce coercions from an element of an abelian category to the set of its pseudoelements
+and from a morphism to the function it induces on pseudoelements.
+
+These coercions must be explicitly enabled via local instances:
+`local attribute [instance] object_to_sort hom_to_fun`
+
+## Implementation notes
+
+It appears that sometimes the coercion from morphisms to functions does not work, i.e.,
+writing `g a` raises a "function expected" error. This error can be fixed by writing
+`(g : X ⟶ Y) a`.
+
+## References
+
+* [F. Borceux, *Handbook of Categorical Algebra 2*][borceux-vol2]
+-/
+
+open category_theory
+open category_theory.limits
+open category_theory.abelian
+open category_theory.preadditive
+
+universes v u
+
+namespace category_theory.abelian
+
+variables {C : Type u} [𝒞 : category.{v} C]
+include 𝒞
+
+/-- Arrows with codomain `P` -/
+def with_codomain (P : C) := Σ Q, Q ⟶ P
```

Comment:
Isn't this just an object of `over P`?

Reply:
You're of course completely right. Fixed.